domain/machine/service/provider.go

```diff
@@ -197,6 +202,70 @@ func (s *ProviderService) GetInstanceTypesFetcher(ctx context.Context) (environs
 	return provider, nil
 }
 
+// UpdateLXDProfiles writes LXD Profiles to LXC for applications on the
+// given machine if the providers supports it. A slice of profile names
+// is returned.
+func (s *ProviderService) UpdateLXDProfiles(ctx context.Context, modelName, machineID string) ([]string, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	provider, err := s.providerGetter(ctx)
+	if err != nil {
+		return nil, errors.Errorf("getting provider: %w", err)
+	}
+
+	if !provider.SupportsLXDProfiles() {
+		s.logger.Tracef(ctx, "LXDProfiler not implemented by environ")
+		return nil, nil
+	}
+
+	profileArgs, err := s.Service.st.GetLXDProfilesForMachine(ctx, machineID)
+	if err != nil {
+		return nil, errors.Errorf("updating LXD profiles in the model for machine %s: %w", machineID, err)
+	}
+
+	var pNames []string
+	for _, arg := range profileArgs {
+		// Lock here, we get a new env for every call to UpdateLXDProfiles().
+		s.lxdProfileMu.Lock()
+		pName := lxdprofile.Name(modelName, arg.ApplicationName, arg.CharmRevision)
+		profile, err := decodeLXDProfile(arg.LXDProfile)
+		if err != nil {
```

Comment:
Also you're not unlocking here correctly.

Reply:
missed that one when I added the decode method.